acme/acme/str_utils.py

```diff
@@ -0,0 +1,65 @@
+"""
```

Comment:
> At the very least, I don't understand why we should use these functions in place of .{de,en}code() in the general case.

`decode` and `encode` are terrible, confusing names for functions that do different type conversions depending on python versions. I had to spend a couple hours reading things and attempting to draw diagrams before I finally figured out which is which where. That's silly. This is much less confusing, and more importantly, matches mypy types. 

> How often do you not know the type of a variable though?

Very often, particularly when trying to use the various crypto libraries.

> Are we worried it might not return a str in Python 2 and bytes in Python 3 for some input? 

No, that's not the problem. The goal is to make it clear to the user what sort of thing you have at the end of doing things. So the code becomes readable and you know what you have, which is not currently the case with or without encode and decode.

`decode` and `encode` are bad and confusing and human beings shouldn't have to use them. the names don't make sense and they don't match what's going on ever.

Reply:
I'm personally not a fan of the idea that you should wrap standard library functions in a language just to provide a better name. Standard library functions are well, standard. Even if the name is bad, people get used to them and learn how to use them as they gain more experience in the language.

We may not have written them much ourselves before now, but I think using these methods is extremely common and it's hard to do anything meaningful in Python 3 without them. We currently have 113 uses of `encode` or `decode` in Certbot.

I personally think writing a custom function when calling a (common) standard library function will do hurts readability more than it helps. Readers have to go check out the definition of your custom functions to see what they're doing only to find out the answer is not much.

> Very often, particularly when trying to use the various crypto libraries.

Unless you're talking about how we have functions that take CSRs and certs which doesn't seem relevant here, I don't think I agree. For instance, there is one use of `isinstance` for determining the string type in all of `acme`.

Reply:
https://bugs.python.org/issue26369

Reply:
I agree with you that `decode` and `encode` could have been designed better. I just think they are very common when writing Python 3 programs and people with experience dealing with I/O in Python 3 will have learned to use these methods before while everyone who works on that part of Certbot would have to learn about our custom methods.

Again, I'm not against you adding these utility methods, I just think always using them when we know the type is a bad idea.

As for that specific issue, I think it luckily doesn't really affect us because our code has to work in Python 2 and 3 where there is no `unicode` type and `str` and `bytes` only have `encode` and `decode` respectively. By having to make our code work in both, I think we avoid the situation of calling `decode` on a `unicode` object or `encode` on a `str` because the equivalent types in Python 3 (`str` and `bytes` respectively) don't have those methods.